DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace DigitalLearningSolutions.Web.ServiceFilter
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Web.Controllers;
+    using DigitalLearningSolutions.Web.Controllers.LearningSolutions;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.Models.Enums;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Filters;
+    using Microsoft.AspNetCore.Mvc.ModelBinding;
+    using Microsoft.FeatureManagement;
+
+    public class ValidateAllowedDlsSubApplication : IActionFilter
+    {
+        private readonly IFeatureManager featureManager;
+        private readonly IEnumerable<DlsSubApplication> validApplications;
+
+        public ValidateAllowedDlsSubApplication(
+            string[] validApplicationNames,
+            IFeatureManager featureManager
+        )
+        {
+            validApplications = validApplicationNames.Select(Enumeration.FromName<DlsSubApplication>);
+            this.featureManager = featureManager;
+        }
+
+        public void OnActionExecuted(ActionExecutedContext context) { }
+
+        public async void OnActionExecuting(ActionExecutingContext context)
+        {
+            var user = context.HttpContext.User;
+            if (!user.Identity.IsAuthenticated)
+            {
+                RedirectToLogin(context);
+                return;
+            }
+
+            var dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
+                .FirstOrDefault(x => x.ParameterType == typeof(DlsSubApplication))?.Name;
+
+            if (!string.IsNullOrEmpty(dlsSubApplicationParameterName) &&
+                HasModelBindingError(context, dlsSubApplicationParameterName!))
+            {
+                SetNotFoundResult(context);
+                return;
+            }
+
+            var application = dlsSubApplicationParameterName == null
+                ? null
+                : GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName);

```

Comment:
If you just go ahead and do this when there's a random string (or empty string) passed in as dlsSubApplication, what happens?

Reply:
And what if the url I go to is <dls>/MyAccount - how does it pick up the `Main` application?

Reply:
When I go to /MyAccount raw, the context contains a 'dlsSubApplicaition' ActionArgument with value 'main'. I admit I wasn't aware this is how it worked before - I thought it just had a null application if there was no route value. I'm not sure where the configuration responsible for this behaviour is.

If the string in the subapp argument is empty or unparseable, it throws.